### DISCUSSION TITLE
DFT field visualization (in adjoint optimizations)

### DISCUSSION THREAD

**Discussion by ghost** [NanoComp/meep#2214] [Q&A] [answered]
I am wondering how to properly add fourier transform monitors in adjoint optimizations, so that they are not removed by the `OptimizationProblem`. It is possible to add the monitor as a `objective_argument` (via `mpa.FourierFields`). This however requires the objective function to accept the transformed fields as an input, which eventually leads to a zero gradient towards this input. Is that problematic, or would that be considered good practice?
One more thing: Is there a convenient way to visualize DFT fields within the "`plot2D`-framework"?
Best Regards
JD

**Comment by smartalecH** [comment, marked as answer]
> This however requires the objective function to accept the transformed fields as an input, which eventually leads to a zero gradient towards this input. Is that problematic, or would that be considered good practice?

This is exactly how we intended the user to arbitrarily pull DFT fields (or other quantities) during an optimization run (I do this regularly). 

If the FOM does not depend on a particular objective argument (or differentiable measurement, as we call them in our [paper](https://doi.org/10.1364/OE.442074)) then no corresponding adjoint source is placed (because the chain rule scale factor is 0, as you suggest). But the specified DFT fields are still accumulated during each forward run. 

> Is there a convenient way to visualize DFT fields within the "plot2D-framework"?

No, not yet. But we welcome contributions!

**Reply by jan-david-fischbach** [reply]
Perfect :) 
> No, not yet. But we welcome contributions!

I might actually look into that at some point. I have always found it very helpful to be able to visualize DFT-fields in Lumerical...
